File: domain/operations.py
"""Operations: deterministic business logic using read and write repository functions.

Handles:
- Resolving driver context (single shipment, ambiguous, no shipments)
- Persisting driver-declared ETAs
- Creating/updating exceptions
- Managing chat threads and messages
- Assembling fresh feasible slot options

All functions call repository.py (no raw SQL).
"""

from typing import Dict, List, Any, Optional
from db import repository as repo
from domain import feasibility


def resolve_driver_operational_context(driver_id: str) -> Dict[str, Any]:
    """
    Resolve the operational context for a trusted driver.
    
    Returns one of:
    1. Single shipment: Complete context dict with shipment, facility, appointment, etc.
    2. Multiple shipments (ambiguous): {'needs_information': True, 'choices': [{'order_reference': ..., 'facility': ...}, ...]}
    3. No active shipments: {'escalate': True, 'message': 'No active shipments assigned.'}
    
    Args:
        driver_id: Trusted driver ID from session
    
    Returns:
        Dict with resolved context or needs_information/escalate markers
    """
    context = repo.get_driver_operational_context(driver_id)
    
    if context is None:
        return {
            'escalate': True,
            'reason': 'NO_ACTIVE_SHIPMENTS',
            'message': f"Driver {driver_id} has no active shipments assigned.",
        }
    
    if context.get('ambiguous'):
        # Multiple active shipments: return choices
        return {
            'needs_information': True,
            'driver_id': driver_id,
            'message': context.get('message'),
            'choices': context.get('choices', []),
        }
    
    # Single shipment: return full context
    return {
        'ready': True,
        'driver_id': driver_id,
        'shipment_id': context['shipment']['shipment_id'],
        'order_reference': context['shipment']['order_reference'],
        'destination_facility': context['facility']['facility_name'],
        'current_status': context['shipment']['current_status'],
        'current_appointment': context['current_appointment'],
        'latest_eta': context['latest_eta'],
        'context': context,  # Full context for downstream use
    }


def persist_driver_declared_eta(
    driver_id: str,
    shipment_id: str,
    declared_eta_ts: str,
    confidence_code: str = 'MEDIUM',
    delay_reason_code: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Persist a driver-declared ETA and create/update exception if delay detected.
    
    Args:
        driver_id: Driver reporting the delay
        shipment_id: Shipment with new ETA
        declared_eta_ts: New ETA (ISO 8601)
        confidence_code: 'HIGH' | 'MEDIUM' | 'LOW'
        delay_reason_code: TRAFFIC, MECHANICAL, WEATHER, etc.
    
    Returns:
        {
            'success': True,
            'eta_update_id': '...',
            'declared_eta_ts': '...',
            'confidence_code': '...',
            'exception_created': bool,
            'exception_id': '...' if exception created,
            'message': str,
        }
    """
    # Persist ETA
    eta_result = repo.persist_driver_eta(
        shipment_id,
        declared_eta_ts,
        confidence_code,
        delay_reason_code,
        reported_by_driver_id=driver_id
    )
    
    # Check if delay detected (declared ETA is later than original)
    shipment = repo.get_shipment(shipment_id)
    original_eta = shipment.get('original_eta_ts') if shipment else None
    
    exception_created = False
    exception_id = None
    exception_type = None
    
    if original_eta and declared_eta_ts > original_eta:
        # Delay detected; create exception
        # Need a thread for the exception
        thread_result = repo.create_or_get_chat_thread(driver_id, shipment_id, 'REPORT_DELAY')
        thread_id = thread_result['thread_id']
        
        exception_type = delay_reason_code if delay_reason_code in (
            'TRAFFIC', 'BREAKDOWN', 'WEATHER'
        ) else 'DELAY'
        
        exc_result = repo.persist_driver_exception(
            driver_id,
            thread_id,
            shipment_id,
            exception_type,
            'NEEDS_INFORMATION',
            description=f"Driver declared new ETA {declared_eta_ts} ({confidence_code} confidence)."
        )
        exception_created = True
        exception_id = exc_result['exception_id']
    
    return {
        'success': True,
        'eta_update_id': eta_result['eta_update_id'],
        'declared_eta_ts': declared_eta_ts,
        'confidence_code': confidence_code,
        'exception_created': exception_created,
        'exception_id': exception_id,
        'message': (
            f"ETA updated to {declared_eta_ts} with {confidence_code} confidence. "
            f"{'Delay exception created.' if exception_created else 'No delay detected.'}"
        ),
    }


def create_or_continue_chat_thread(
    driver_id: str,
    shipment_id: Optional[str] = None,
    thread_reason: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Create a new chat thread for a driver or continue an existing open thread.
    
    Args:
        driver_id: Driver starting/continuing conversation
        shipment_id: Optional; which shipment (None if ambiguous/unresolved)
        thread_reason: REPORT_DELAY | EARLY_ARRIVAL | CHECK_STATUS | ASK_SLOT_OPTIONS
    
    Returns:
        {
            'thread_id': '...',
            'created': bool,
            'message': str,
        }
    """
    return repo.create_or_get_chat_thread(driver_id, shipment_id, thread_reason)


def persist_incoming_driver_message(
    thread_id: str,
    driver_id: str,
    message_text: str,
    external_message_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Persist an incoming message from driver.
    
    Detects and marks duplicates (retries from SMS/messaging gateway).
    
    Args:
        thread_id: Chat thread
        driver_id: Driver ID
        message_text: Message content
        external_message_id: Gateway message ID (for dedup)
    
    Returns:
        {
            'chat_message_id': '...',
            'is_duplicate': bool,
            'message': str,
        }
    """
    result = repo.persist_chat_message(
        thread_id,
        'DRIVER',
        driver_id,
        message_text,
        external_message_id=external_message_id
    )
    
    return {
        'chat_message_id': result['chat_message_id'],
        'is_duplicate': result['is_duplicate'],
        'message': result['message'],
    }


def persist_outgoing_agent_message(
    thread_id: str,
    message_text: str,
    external_message_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Persist an outgoing message from agent.
    
    Args:
        thread_id: Chat thread
        message_text: Agent response
        external_message_id: Optional gateway reference
    
    Returns:
        {
            'chat_message_id': '...',
            'message': str,
        }
    """
    result = repo.persist_chat_message(
        thread_id,
        'AGENT',
        'agent',
        message_text,
        external_message_id=external_message_id
    )
    
    return {
        'chat_message_id': result['chat_message_id'],
        'message': result['message'],
    }


def detect_duplicate_incoming_message(
    external_message_id: Optional[str],
) -> bool:
    """
    Check if an incoming message is a duplicate (retry).
    
    Args:
        external_message_id: Message ID from gateway
    
    Returns:
        True if duplicate, False otherwise
    """
    if not external_message_id:
        return False
    
    existing = repo.detect_duplicate_message(external_message_id, None)
    return existing is not None


def assemble_fresh_feasible_options(
    driver_id: str,
    shipment_id: str,
    max_options: int = 5,
) -> Dict[str, Any]:
    """
    Assemble fresh set of feasible slots for a shipment.
    
    Calls feasibility.py to evaluate each candidate slot.
    
    Args:
        driver_id: Driver requesting options (for validation)
        shipment_id: Shipment needing a slot
        max_options: Maximum options to return
    
    Returns:
        {
            'shipment_id': shipment_id,
            'facility_id': facility_id,
            'total_available': int,
            'total_feasible': int,
            'feasible': [
                {
                    'slot_id': '...',
                    'dock_code': '...',
                    'slot_start_ts': '...',
                    'slot_end_ts': '...',
                    'recommendation': 'BOOK' | 'NEEDS_APPROVAL',
                },
                ...
            ],
            'needs_approval_count': int,
            'escalation_reason': str or None,
        }
    """
    shipment = repo.get_shipment(shipment_id)
    if not shipment:
        return {
            'error': 'SHIPMENT_NOT_FOUND',
            'message': f'Shipment {shipment_id} not found.',
        }
    
    facility_id = shipment['destination_facility_id']
    dock_type = shipment.get('required_dock_type')
    
    # Use feasibility to evaluate slots
    return feasibility.assemble_feasible_slot_options(
        shipment_id,
        facility_id,
        dock_type=dock_type,
        max_options=max_options
    )


def create_or_update_exception(
    driver_id: str,
    thread_id: str,
    shipment_id: Optional[str],
    exception_type: str,
    exception_status: str = 'NEEDS_INFORMATION',
    description: str = '',
) -> Dict[str, Any]:
    """
    Create an exception record.
    
    Args:
        driver_id: Driver reporting issue
        thread_id: Chat context
        shipment_id: Optional affected shipment
        exception_type: 'DELAY' | 'BREAKDOWN' | 'TRAFFIC' | 'WEATHER' | 'EARLY_ARRIVAL' | 'DOCK_UNAVAILABLE' | 'UNKNOWN'
        exception_status: 'OPEN' | 'NEEDS_INFORMATION' | 'SLOT_OPTIONS_SHARED' | 'WAITING_CONFIRMATION' | 'RESOLVED' | 'ESCALATED' | 'DUPLICATE' | 'CANCELLED'
        description: Human-readable description
    
    Returns:
        {
            'exception_id': '...',
            'exception_type': '...',
            'exception_status': '...',
            'message': str,
        }
    """
    return repo.persist_driver_exception(
        driver_id,
        thread_id,
        shipment_id,
        exception_type,
        exception_status,
        description=description
    )

File: src/domain/operations.ts
import { db } from '../db/dataStore';

export function resolveDriverOperationalContext(driverId: string) {
  const context = db.getDriverOperationalContext(driverId);
  if (!context) {
    return {
      status: 'escalate',
      driver_id: driverId,
      message: 'No active shipments found for this driver.',
    };
  }

  if (context.isPendingApproval) {
    return {
      status: 'pending_approval',
      driver_id: driverId,
      driver_name: context.driver_name,
      approval_status: 'PENDING',
      vehicle_registration: context.vehicle_registration,
      registered_at: context.registered_at,
      message: 'Driver registration is currently being processed by the facility coordinator.',
    };
  }

  if (context.ambiguous) {
    return {
      status: 'needs_information',
      driver_id: driverId,
      choices: context.choices,
      choice_count: context.choice_count,
      message: context.message,
    };
  }

  return {
    status: 'ready',
    driver_id: driverId,
    shipment_id: context.shipment.shipment_id,
    order_reference: context.shipment.order_reference,
    carrier_name: context.carrier?.carrier_name,
    vehicle_registration: context.vehicle?.registration_number,
    facility_id: context.facility?.facility_id,
    facility_name: context.facility?.facility_name,
    required_dock_type: context.shipment.required_dock_type,
    effective_eta: context.latest_eta?.effective_eta_ts,
    eta_confidence: context.latest_eta?.eta_confidence,
    current_appointment: context.current_appointment ? {
      appointment_id: context.current_appointment.appointment_id,
      slot_start_ts: context.current_appointment.slot?.slot_start_ts,
      slot_end_ts: context.current_appointment.slot?.slot_end_ts,
      dock_code: context.current_appointment.dock_code,
      status: context.current_appointment.appointment_status,
      is_warehouse_confirmed: context.current_appointment.is_warehouse_confirmed,
      warehouse_confirmation_ref: context.current_appointment.warehouse_confirmation_ref,
    } : null,
    active_exceptions: context.dock_status_events?.length || 0,
    checkin_state: context.facility_checkin?.queue_state || null,
  };
}

export function selectDriverShipmentByOrderReference(driverId: string, orderRef: string) {
  const shipmentId = db.getShipmentIdByOrderReferenceForDriver(driverId, orderRef);
  if (!shipmentId) {
    const activeShipments = db.getDriverActiveShipments(driverId);
    return {
      status: 'not_found',
      driver_id: driverId,
      order_reference: orderRef,
      message: `No active shipment with order reference "${orderRef}" found for driver ${driverId}.`,
      valid_choices: activeShipments.map(s => s.order_reference),
    };
  }

  const shipment = db.getShipment(shipmentId)!;
  const facility = db.getFacility(shipment.destination_facility_id);
  const latestEta = db.getLatestETA(shipmentId);
  const currentAppt = db.getCurrentAppointment(shipmentId);

  return {
    status: 'ready',
    driver_id: driverId,
    shipment_id: shipment.shipment_id,
    order_reference: shipment.order_reference,
    facility_id: shipment.destination_facility_id,
    facility_name: facility ? facility.facility_name : shipment.destination_facility_id,
    required_dock_type: shipment.required_dock_type,
    effective_eta: latestEta?.effective_eta_ts,
    eta_confidence: latestEta?.eta_confidence,
    current_appointment: currentAppt ? {
      appointment_id: currentAppt.appointment_id,
      slot_start_ts: currentAppt.slot?.slot_start_ts,
      slot_end_ts: currentAppt.slot?.slot_end_ts,
      dock_code: currentAppt.dock_code,
      status: currentAppt.appointment_status,
      is_warehouse_confirmed: currentAppt.is_warehouse_confirmed,
      warehouse_confirmation_ref: currentAppt.warehouse_confirmation_ref,
    } : null,
  };
}

export function persistDriverDeclaredETA(
  driverId: string,
  shipmentId: string,
  declaredEtaTs: string,
  confidenceCode: 'LOW' | 'MEDIUM' | 'HIGH',
  delayReasonCode?: string | null
) {
  const thread = db.createOrGetChatThread(driverId, shipmentId, 'REPORT_DELAY');
  const etaResult = db.persistDriverETA(shipmentId, declaredEtaTs, confidenceCode, delayReasonCode, driverId);

  const exceptionResult = db.persistDriverException(
    driverId,
    thread.thread_id,
    shipmentId,
    'DELAY',
    'NEEDS_INFORMATION',
    `Driver declared ETA ${declaredEtaTs} (confidence: ${confidenceCode}, reason: ${delayReasonCode || 'Not provided'})`
  );

  return {
    status: 'ok',
    shipment_id: shipmentId,
    declared_eta_ts: declaredEtaTs,
    confidence_code: confidenceCode,
    delay_reason_code: delayReasonCode || null,
    eta_update_id: etaResult.eta_update_id,
    exception_id: exceptionResult.exception_id,
    thread_id: thread.thread_id,
    message: `ETA recorded: ${declaredEtaTs}. Exception registered with status NEEDS_INFORMATION.`,
  };
}

export function getExceptionOrAppointmentStatus(driverId: string, shipmentId?: string | null) {
  let targetShipmentId = shipmentId;
  if (!targetShipmentId) {
    const active = db.getDriverActiveShipments(driverId);
    if (active.length === 1) {
      targetShipmentId = active[0].shipment_id;
    }
  }

  if (!targetShipmentId) {
    return {
      status: 'needs_information',
      message: 'Could not uniquely identify the shipment to check status for.',
    };
  }

  const appt = db.getCurrentAppointment(targetShipmentId);
  const eta = db.getLatestETA(targetShipmentId);
  const shipment = db.getShipment(targetShipmentId);

  return {
    status: 'ok',
    shipment_id: targetShipmentId,
    order_reference: shipment?.order_reference,
    current_status: shipment?.current_status,
    effective_eta: eta?.effective_eta_ts,
    eta_confidence: eta?.eta_confidence,
    appointment: appt ? {
      appointment_id: appt.appointment_id,
      slot_id: appt.slot_id,
      slot_start_ts: appt.slot?.slot_start_ts,
      slot_end_ts: appt.slot?.slot_end_ts,
      dock_code: appt.dock_code,
      status: appt.appointment_status,
      is_warehouse_confirmed: appt.is_warehouse_confirmed,
      warehouse_confirmation_ref: appt.warehouse_confirmation_ref,
      booked_at: appt.booked_at,
    } : null,
  };
}
